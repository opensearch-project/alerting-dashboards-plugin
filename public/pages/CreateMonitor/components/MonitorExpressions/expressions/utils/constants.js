/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
  METRICS: 'METRICS',
  GROUP_BY: 'GROUP_BY',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: {
    text: 'is',
    value: 'is',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT: {
    text: 'is not',
    value: 'is_not',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NULL: {
    text: 'is null',
    value: 'is_null',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT_NULL: {
    text: 'is not null',
    value: 'is_not_null',
    dataTypes: ['number', 'text', 'keyword'],
  },
  IS_GREATER: {
    text: 'is greater than',
    value: 'is_greater',
    dataTypes: ['number'],
  },
  IS_GREATER_EQUAL: {
    text: 'is greater than equal',
    value: 'is_greater_equal',
    dataTypes: ['number'],
  },
  IS_LESS: {
    text: 'is less than',
    value: 'is_less',
    dataTypes: ['number'],
  },
  IS_LESS_EQUAL: {
    text: 'is less than equal',
    value: 'is_less_equal',
    dataTypes: ['number'],
  },
  STARTS_WITH: {
    text: 'starts with',
    value: 'starts_with',
    dataTypes: ['text', 'keyword'],
  },
  ENDS_WITH: {
    text: 'ends with',
    value: 'ends_with',
    dataTypes: ['text', 'keyword'],
  },
  CONTAINS: {
    text: 'contains',
    value: 'contains',
    dataTypes: ['text', 'keyword'],
  },
  DOES_NOT_CONTAINS: {
    text: 'does not contain',
    value: 'does_not_contains',
    dataTypes: ['text'],
  },
  IN_RANGE: {
    text: 'is in range',
    value: 'in_range',
    dataTypes: ['number'],
  },
  NOT_IN_RANGE: {
    text: 'is not in range',
    value: 'not_in_range',
    dataTypes: ['number'],
  },
};

export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const GROUP_BY_ERROR = 'Must specify at least 1 group by expression.';
export const QUERY_TYPE_GROUP_BY_ERROR = 'Can have a maximum of 1 group by selections.';

export const QUERY_TYPE_METRIC_ERROR = 'Can have a maximum of 1 metric selections.';

export const MAX_NUM_QUERY_LEVEL_GROUP_BYS = 1;
export const MAX_NUM_BUCKET_LEVEL_GROUP_BYS = 2;
export const MAX_NUM_WHERE_EXPRESSION = 1;
export const WHERE_FILTER_ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];
