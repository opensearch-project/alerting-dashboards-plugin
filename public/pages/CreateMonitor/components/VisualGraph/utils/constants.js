/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const SIZE_RANGE = [3, 3];
export const LINE_STYLES = { strokeWidth: 1 };
export const ANNOTATION_STYLES = { strokeWidth: 1, stroke: 'red' };
export const HINT_STYLES = {
  background: '#3a3a48',
  borderRadius: '5px',
  padding: '7px 10px',
  fontSize: '12px',
  fontWeight: 'bold',
  color: '#fff',
  textAlign: 'left',
  whiteSpace: 'nowrap',
  boxShadow: '0 2px 4px rgba(0,0,0,.5)',
};

// Don't want values to reach top of graph so increase by a buffer % so we always have a yDomain with a slightly higher max
export const Y_DOMAIN_BUFFER = 1.4; // 40%

export const X_DOMAIN_BUFFER = 0.2; // 40%

// Size of circle for each point on graph
export const DEFAULT_MARK_SIZE = 3;

export const BAR_PERCENTAGE = 0.015;

export const BAR_KEY_COUNT = 3;
