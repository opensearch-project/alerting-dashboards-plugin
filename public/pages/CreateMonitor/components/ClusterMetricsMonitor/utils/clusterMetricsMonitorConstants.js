/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { DEFAULT_EMPTY_DATA } from '../../../../../utils/constants';

export const URL_DEFAULT_PREFIX = 'http://localhost:9200';
export const API_PATH_REQUIRED_PLACEHOLDER_TEXT = 'Select an API.';
export const EMPTY_PATH_PARAMS_TEXT = 'Enter remaining path components and path parameters';
export const GET_API_TYPE_DEBUG_TEXT =
  'Cannot determine ApiType in clusterMetricsMonitorHelpers::getSelectedApiType.';
export const ILLEGAL_PATH_PARAMETER_CHARACTERS = [
  ':',
  '"',
  '+',
  '\\',
  '|',
  '?',
  '#',
  '>',
  '<',
  ' ',
];
export const NO_PATH_PARAMS_PLACEHOLDER_TEXT = 'No path parameter options';
export const PATH_PARAMETER_ILLEGAL_CHARACTER_TEXT = `The provided path parameters contain invalid characters or spaces. Please omit: ${ILLEGAL_PATH_PARAMETER_CHARACTERS.join(
  ' '
)}`;
export const PATH_PARAMETERS_REQUIRED_TEXT = 'Path parameters are required for this API.';
export const REST_API_REFERENCE = 'https://opensearch.org/docs/latest/opensearch/rest-api/index/';
export const DEFAULT_CLUSTER_METRICS_SCRIPT = {
  lang: 'painless',
  source: 'ctx.results[0] != null',
};

export const CLUSTER_METRICS_CROSS_CLUSTER_ALERT_TABLE_COLUMN = {
  field: 'clusters',
  name: 'Triggered clusters',
  sortable: true,
  truncateText: true,
  render: (clusters = [DEFAULT_EMPTY_DATA]) => _.sortBy(clusters).join(', '),
};

export const API_TYPES = {
  CLUSTER_HEALTH: {
    type: 'CLUSTER_HEALTH',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cluster-health/',
    exampleText: 'indexAlias1,indexAlias2...',
    label: 'Cluster health',
    paths: {
      withPathParams: '_cluster/health',
      withoutPathParams: '_cluster/health',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].status != "green"',
    },
  },
  CLUSTER_STATS: {
    type: 'CLUSTER_STATS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cluster-stats/',
    exampleText: 'nodeFilter1,nodeFilter2...',
    label: 'Cluster stats',
    paths: {
      withPathParams: '_cluster/stats/nodes',
      withoutPathParams: '_cluster/stats',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].indices.count <= 0',
    },
  },
  CLUSTER_SETTINGS: {
    type: 'CLUSTER_SETTINGS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cluster-settings/',
    exampleText: undefined,
    label: 'Cluster settings',
    paths: {
      withPathParams: undefined,
      withoutPathParams: '_cluster/settings',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].transient != null',
    },
  },
  NODES_STATS: {
    type: 'NODES_STATS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/popular-api/#get-node-statistics',
    exampleText: undefined,
    label: 'Nodes stats',
    paths: {
      withPathParams: undefined,
      withoutPathParams: '_nodes/stats',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].nodes.NODE_ID.jvm.mem.heap_used_percent > 60',
    },
  },
  CAT_INDICES: {
    type: 'CAT_INDICES',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-indices/',
    exampleText: 'index1,index2...',
    label: 'List indices',
    paths: {
      withPathParams: '_cat/indices',
      withoutPathParams: '_cat/indices',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: `for (int i = 0; i < ctx.results[0].indices.size(); ++i)
        if (ctx.results[0].indices[i].health != "green") return true`,
    },
  },
  CAT_PENDING_TASKS: {
    type: 'CAT_PENDING_TASKS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-pending-tasks/',
    exampleText: undefined,
    label: 'List pending tasks',
    paths: {
      withPathParams: undefined,
      withoutPathParams: '_cat/pending_tasks',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].tasks.size() >= 0',
    },
  },
  CAT_RECOVERY: {
    type: 'CAT_RECOVERY',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-recovery/',
    exampleText: 'index1,index2...',
    label: 'List index and shard recoveries',
    paths: {
      withPathParams: '_cat/recovery',
      withoutPathParams: '_cat/recovery',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].INDEX_NAME.shards.length <= 0',
    },
  },
  CAT_SHARDS: {
    type: 'CAT_SHARDS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-shards/',
    exampleText: 'index1,index2...',
    label: 'List shards',
    paths: {
      withPathParams: '_cat/shards',
      withoutPathParams: '_cat/shards',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: `for (int i = 0; i < ctx.results[0].shards.size(); ++i)
        if (ctx.results[0].shards[i]["unassigned.for"] != null) return true`,
    },
  },
  CAT_SNAPSHOTS: {
    type: 'CAT_SNAPSHOTS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-snapshots/',
    exampleText: 'repositoryName',
    label: 'List snapshots',
    paths: {
      withPathParams: '_cat/snapshots',
      withoutPathParams: undefined,
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].SNAPSHOT_ID.status == "FAILED"',
    },
  },
  CAT_TASKS: {
    type: 'CAT_TASKS',
    documentation: 'https://opensearch.org/docs/latest/opensearch/rest-api/cat/cat-tasks/',
    exampleText: undefined,
    label: 'List tasks',
    paths: {
      withPathParams: undefined,
      withoutPathParams: '_cat/tasks',
    },
    get prependText() {
      return this.paths.withPathParams || this.paths.withoutPathParams;
    },
    appendText: '',
    defaultCondition: {
      ...DEFAULT_CLUSTER_METRICS_SCRIPT,
      source: 'ctx.results[0].tasks.length > 0',
    },
  },
};
