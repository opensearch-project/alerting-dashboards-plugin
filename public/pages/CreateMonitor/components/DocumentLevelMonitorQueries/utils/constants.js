/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { OPERATORS_MAP } from '../../MonitorExpressions/expressions/utils/constants';

export const DOC_LEVEL_INPUT_FIELD = 'doc_level_input';

/**
 * A list of the operators currently supported for defining queries through the UI.
 */
export const SUPPORTED_DOC_LEVEL_QUERY_OPERATORS = [
  OPERATORS_MAP.IS.value,
  OPERATORS_MAP.IS_NOT.value,
  OPERATORS_MAP.IS_GREATER.value,
  OPERATORS_MAP.IS_GREATER_EQUAL.value,
  OPERATORS_MAP.IS_LESS.value,
  OPERATORS_MAP.IS_LESS_EQUAL.value,
];

/**
 * Initial implementation of document level monitors used these feature-specific operator values.
 * Refactored document level monitors assets to reuse the OPERATORS_MAP to reduce duplicate code
 * as new query types become supported.
 */
export const LEGACY_QUERY_OPERATORS = {
  IS: { text: 'is', value: '==' },
  IS_NOT: { text: 'is not', value: '!=' },
};

/**
 * These patterns delineate the field being queries from the query value
 * when using query string query syntax.
 */
export const QUERY_STRING_QUERY_OPERATORS = {
  [OPERATORS_MAP.IS_GREATER.value]: ':>',
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: ':>=',
  [OPERATORS_MAP.IS_LESS.value]: ':<',
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: ':<=',
};

/**
 * Similar to OPERATORS_QUERY_MAP, this const contains the query patterns
 * for the SUPPORTED_DOC_LEVEL_QUERY_OPERATORS.
 */
export const DOC_LEVEL_QUERY_MAP = {
  [LEGACY_QUERY_OPERATORS.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [LEGACY_QUERY_OPERATORS.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [OPERATORS_MAP.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS_GREATER.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER.value]}${query}`,
  },
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER_EQUAL.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS_EQUAL.value]}${query}`,
  },
};
