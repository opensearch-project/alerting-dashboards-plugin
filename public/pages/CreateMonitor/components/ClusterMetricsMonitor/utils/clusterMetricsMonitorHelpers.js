/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { formikToClusterMetricsInput } from '../../../containers/CreateMonitor/utils/formikToMonitor';
import {
  API_TYPES,
  GET_API_TYPE_DEBUG_TEXT,
  ILLEGAL_PATH_PARAMETER_CHARACTERS,
  NO_PATH_PARAMS_PLACEHOLDER_TEXT,
  PATH_PARAMETER_ILLEGAL_CHARACTER_TEXT,
  PATH_PARAMETERS_REQUIRED_TEXT,
} from './clusterMetricsMonitorConstants';
import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';

export function buildClusterMetricsRequest(values) {
  return _.get(formikToClusterMetricsInput(values), 'uri');
}

export const canExecuteClusterMetricsMonitor = (uri = {}) => {
  const {
    api_type = FORMIK_INITIAL_VALUES.uri.api_type,
    path_params = FORMIK_INITIAL_VALUES.uri.path_params,
  } = uri;
  if (_.isEmpty(api_type)) return false;
  const requiresPathParams = _.isEmpty(_.get(API_TYPES, `${api_type}.paths.withoutPathParams`));
  const hasPathParams = !_.isEmpty(path_params);
  if (pathParamsContainIllegalCharacters(path_params)) return false;
  return requiresPathParams ? hasPathParams : true;
};

export const pathParamsContainIllegalCharacters = (pathParams) => {
  if (_.isEmpty(pathParams)) return false;
  const foundIllegalCharacters = ILLEGAL_PATH_PARAMETER_CHARACTERS.find((illegalCharacter) =>
    _.includes(pathParams, illegalCharacter)
  );
  return !_.isEmpty(foundIllegalCharacters);
};

export const getApiPath = (hasPathParams = false, apiType) => {
  let path = hasPathParams
    ? _.get(
        API_TYPES,
        `${apiType}.paths.withPathParams`,
        _.get(API_TYPES, `${apiType}.paths.withoutPathParams`)
      )
    : _.get(API_TYPES, `${apiType}.paths.withoutPathParams`);
  return path || FORMIK_INITIAL_VALUES.uri.path;
};

export const getApiTypesRequiringPathParams = () => {
  const apiList = [];
  _.keys(API_TYPES).forEach((api) => {
    const withoutPathParams = _.get(API_TYPES, `${api}.paths.withoutPathParams`, '');
    if (_.isEmpty(withoutPathParams))
      apiList.push({
        value: _.get(API_TYPES, `${api}.type`),
        label: _.get(API_TYPES, `${api}.label`),
      });
  });
  return apiList;
};

export const getExamplePathParams = (apiType) => {
  if (_.isEmpty(apiType)) return NO_PATH_PARAMS_PLACEHOLDER_TEXT;
  let exampleText = _.get(API_TYPES, `${apiType}.exampleText`, '');
  _.isEmpty(exampleText)
    ? (exampleText = NO_PATH_PARAMS_PLACEHOLDER_TEXT)
    : (exampleText = `e.g., ${exampleText}`);
  return exampleText;
};

export const getApiType = (uri) => {
  let apiType = '';
  // Trim '/' characters from the beginning and end of the path
  const path = _.get(uri, 'path')?.replace(/^\/+|\/+$/g, '');
  if (_.isEmpty(path)) return apiType;
  _.keys(API_TYPES).forEach((apiTypeKey) => {
    const withPathParams = _.get(API_TYPES, `${apiTypeKey}.paths.withPathParams`);
    const withoutPathParams = _.get(API_TYPES, `${apiTypeKey}.paths.withoutPathParams`);
    if (path === withPathParams || path === withoutPathParams) apiType = apiTypeKey;
  });
  if (_.isEmpty(apiType)) console.debug(GET_API_TYPE_DEBUG_TEXT);
  return apiType;
};

export const isInvalidApiPathParameter = (field, hidePathParams, pathParams, requirePathParams) => {
  if (hidePathParams) return false;
  const pathParamsTouched = _.get(field, 'touched.uri.path_params', false);
  if (pathParamsTouched) {
    if (requirePathParams && _.isEmpty(pathParams)) return true;
    return pathParamsContainIllegalCharacters(pathParams);
  } else return pathParamsTouched;
};

export const validateApiPathParameter = (field, hidePathParams, pathParams, requirePathParams) => {
  if (hidePathParams) return NO_PATH_PARAMS_PLACEHOLDER_TEXT;
  const pathParamsTouched = _.get(field, 'touched.uri.path_params', false);
  if (requirePathParams && pathParamsTouched && _.isEmpty(pathParams))
    return PATH_PARAMETERS_REQUIRED_TEXT;
  if (isInvalidApiPathParameter(field, hidePathParams, pathParams, requirePathParams))
    return PATH_PARAMETER_ILLEGAL_CHARACTER_TEXT;
};
