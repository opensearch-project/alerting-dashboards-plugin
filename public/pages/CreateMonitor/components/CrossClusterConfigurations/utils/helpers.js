/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { DEFAULT_EMPTY_DATA, MONITOR_TYPE } from '../../../../../utils/constants';

export const getLocalClusterName = async (httpClient) => {
  let localClusterName = DEFAULT_EMPTY_DATA;
  try {
    const response = await httpClient.get('../api/alerting/_health');
    if (response.ok) {
      localClusterName = response.resp[0]?.cluster;
    } else {
      console.log('Error getting local cluster name:', response);
    }
  } catch (e) {
    console.error(e);
  }
  return localClusterName;
};

export const getDataSources = (monitor, localClusterName) => {
  const monitorType = _.get(
    monitor,
    'monitor_type',
    _.get(monitor, 'ui_metadata.monitor_type', MONITOR_TYPE.QUERY_LEVEL)
  );
  let dataSources;
  switch (monitorType) {
    case MONITOR_TYPE.CLUSTER_METRICS:
      dataSources = _.get(monitor, 'inputs.0.uri.clusters');
      // To preserve functionality of legacy monitors, cluster metrics monitors run on the
      // local cluster by default if no clusters are specified in the monitor configuration.
      if (_.isEmpty(dataSources)) dataSources = [localClusterName || DEFAULT_EMPTY_DATA];
      break;
    case MONITOR_TYPE.DOC_LEVEL:
      dataSources = _.get(monitor, 'inputs.0.doc_level_input.indices', [DEFAULT_EMPTY_DATA]);
      break;
    default:
      dataSources = _.get(monitor, 'inputs.0.search.indices', [DEFAULT_EMPTY_DATA]);
  }
  dataSources = _.sortBy(dataSources);
  return dataSources;
};
