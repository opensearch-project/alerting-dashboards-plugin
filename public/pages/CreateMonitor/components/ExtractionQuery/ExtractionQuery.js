/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiCodeEditor, EuiFlexGroup, EuiFlexItem, EuiCompressedFormRow } from '@elastic/eui';
import 'brace/mode/json';
import 'brace/mode/plain_text';
import 'brace/snippets/javascript';
import 'brace/ext/language_tools';

import { FormikCodeEditor } from '../../../../components/FormControls';
import { isInvalid, hasError, validateExtractionQuery } from '../../../../utils/validate';

const ExtractionQuery = ({ isDarkMode, response }) => (
  <EuiFlexGroup>
    <EuiFlexItem>
      <FormikCodeEditor
        name="query"
        formRow
        fieldProps={{ validate: validateExtractionQuery }}
        rowProps={{
          label: 'Define extraction query',
          fullWidth: true,
          isInvalid,
          error: hasError,
        }}
        inputProps={{
          mode: 'json',
          width: '100%',
          height: '500px',
          theme: isDarkMode ? 'sense-dark' : 'github',
          onChange: (query, field, form) => {
            form.setFieldValue('query', query);
          },
          onBlur: (e, field, form) => {
            form.setFieldTouched('query', true);
          },
          'data-test-subj': 'extractionQueryCodeEditor',
        }}
      />
    </EuiFlexItem>
    <EuiFlexItem>
      <EuiCompressedFormRow label="Extraction query response" fullWidth>
        <EuiCodeEditor
          mode="json"
          theme={isDarkMode ? 'sense-dark' : 'github'}
          width="100%"
          height="500px"
          value={response}
          readOnly
        />
      </EuiCompressedFormRow>
    </EuiFlexItem>
  </EuiFlexGroup>
);

export default ExtractionQuery;
