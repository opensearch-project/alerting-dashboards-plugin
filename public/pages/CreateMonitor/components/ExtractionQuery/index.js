/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ExtractionQuery from './ExtractionQuery';

export default ExtractionQuery;
