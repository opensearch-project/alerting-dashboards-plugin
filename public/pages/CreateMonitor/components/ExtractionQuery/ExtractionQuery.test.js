/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from 'enzyme';

import ExtractionQuery from './ExtractionQuery';
import { Formik } from 'formik';
import { FORMIK_INITIAL_VALUES } from '../../containers/CreateMonitor/utils/constants';

describe('ExtractionQuery', () => {
  test('renders', () => {
    const component = (
      <Formik initialValues={FORMIK_INITIAL_VALUES}>
        {() => <ExtractionQuery response={JSON.stringify({ hits: 10 })} />}
      </Formik>
    );

    expect(render(component)).toMatchSnapshot();
  });
});
