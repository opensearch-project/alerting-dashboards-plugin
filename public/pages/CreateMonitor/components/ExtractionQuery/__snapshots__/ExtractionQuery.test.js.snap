// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtractionQuery renders 1`] = `
<div
  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
>
  <div
    class="euiFlexItem"
  >
    <div
      class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
      id="query-form-row-row"
    >
      <div
        class="euiFormRow__labelWrapper"
      >
        <label
          aria-invalid="false"
          class="euiFormLabel euiFormRow__label"
          for="query-form-row"
        >
          Define extraction query
        </label>
      </div>
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiCodeEditorWrapper"
          data-test-subj="extractionQueryCodeEditor"
          style="width:100%;height:500px"
        >
          <button
            class="euiCodeEditorKeyboardHint"
            data-test-subj="codeEditorHint"
            id="generated-id"
          >
            <p
              class="euiText"
            >
              Press Enter to start editing.
            </p>
            <p
              class="euiText"
            >
              When you're done, press Escape to stop editing.
            </p>
          </button>
          <div
            id="generated-id"
            style="width:100%;height:500px"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem"
  >
    <div
      class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
      id="generated-id-row"
    >
      <div
        class="euiFormRow__labelWrapper"
      >
        <label
          class="euiFormLabel euiFormRow__label"
          for="generated-id"
        >
          Extraction query response
        </label>
      </div>
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiCodeEditorWrapper"
          data-test-subj="codeEditorContainer"
          style="width:100%;height:500px"
        >
          <button
            class="euiCodeEditorKeyboardHint"
            data-test-subj="codeEditorHint"
            id="generated-id"
          >
            <p
              class="euiText"
            >
              Press Enter to start editing.
            </p>
            <p
              class="euiText"
            >
              When you're done, press Escape to stop editing.
            </p>
          </button>
          <div
            id="generated-id"
            style="width:100%;height:500px"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
